src/thor/route_action.cc

```diff
@@ -446,6 +454,14 @@ void thor_worker_t::path_arrive_by(Api& api, const std::string& costing) {
         TripLegBuilder::Build(options, controller, *reader, mode_costing, path.begin(), path.end(),
                               *origin, *destination, leg, algorithms, interrupt, edge_trimming,
                               intermediates);
+
+        // advance the time for the next destination (i.e. algo origin) by the waiting_secs
+        // of this origin (i.e. algo destination)
+        if (origin->waiting_secs()) {
+          // no need for timezones, we'll wait at this location
+          auto origin_dt = offset_date(*reader, origin->date_time(), {}, -origin->waiting_secs(), {});
+          origin->set_date_time(origin_dt);
+        }
```

Comment:
Still inside the same loop through all alternates, just further down after we built this path's `TripLeg` . After the first leg was built, this will advance the `origin`'s time by `waiting` secs, because the `origin` will become the next leg's destination. And in the reverse A* `destination` & `origin` are flipped, so this makes sure that the next reverse expansion starts with the correct time.